src/modules/job-info/watch.c

```diff
@@ -258,16 +274,29 @@ static void watch_continuation (flux_future_t *f, void *arg)
     if (flux_kvs_lookup_get (f, &s) < 0) {
         if (errno != ENOENT && errno != ENODATA && errno != ENOTSUP)
             flux_log_error (ctx->h, "%s: flux_kvs_lookup_get", __FUNCTION__);
-        if (errno == ENODATA && w->kvs_watch_canceled && !w->cancel)
-            goto cleanup;
+        if (errno == ENODATA) {
+            if (w->kvs_watch_canceled && !w->cancel)
+                goto cleanup;
+            /* N.B. if guest in main, we just read / streamed the
+             * entire contents from KVS and that's it.  So add
+             * sentinel before ENODATA at the end.
+             */
+            if (w->flags & FLUX_JOB_EVENT_WATCH_INITIAL_SENTINEL
+                && w->guest_in_main
+                && !w->initial_sentinel_sent)
```

Comment:
I think the test of `!w->initial_sentinel_sent` is redundant with the one in `send_initial_sentinel()`